### PR TITLE
Enables LRP to react on pod readiness.

### LINKED ISSUE #13168: Health-check backend pods when LRP is active
Per offliine discussion with @aditighag, periodically health-checking backends picked by active LRP is desired to ensure cluster functionality.

A few concerns:
- If all backends become unhealthy, should we remove/circumvent LRP? What's the channel of notification to users that it is no longer in effect?
- Should we drop LRP entirely if all backends are unhealthy for a prolonged time?

### PR DESCRIPTION
**Pull request by Weil0ng**
Currently local redirect service won't react if backend pod becomes
unhealty.

This adds logic that checks the pod status within watcher and update LRP
configs accordingly.

To achieve the above goal, PodConditions are added into the slim apis to
allow for watching the "ready" condition of pods.

This PR also adds logic in LRP manager to restore service when no
backend is healthy.

Signed-off-by: Weilong Cui <cuiwl@google.com>

Fixes: #13168

### CODE REVIEW

File: pkg/k8s/slim/k8s/apis/core/v1/types.go [hunk 1]
```diff
@@ -120,6 +120,64 @@ const (
...
+
+// These are valid condition statuses. "ConditionTrue" means a resource is in the condition.
+// "ConditionFalse" means a resource is not in the condition. "ConditionUnknown" means kubernetes
+// can't decide if a resource is in the condition or not. In the future, we could add other
+// intermediate conditions, e.g. ConditionDegraded.
+const (
+        ConditionTrue    ConditionStatus = "True"
+        ConditionFalse   ConditionStatus = "False"
+        ConditionUnknown ConditionStatus = "Unknown"
+)
```
Comment (aanm) [inline pkg/k8s/slim/k8s/apis/core/v1/types.go]: This needs to be between lines 99 and 100

File: pkg/k8s/watchers/pod.go [hunk 2]
```diff
@@ -276,7 +277,24 @@ func (k *K8sWatcher) updateK8sPodV1(oldK8sPod, newK8sPod *slim_corev1.Pod) error
...
+	for _, cond := range oldK8sPod.Status.Conditions {
+		if cond.Type == slim_corev1.PodReady {
+			oldPodReady = cond.Status
+		}
+	}
+	for _, cond := range newK8sPod.Status.Conditions {
+		if cond.Type == slim_corev1.PodReady {
+			newPodReady = cond.Status
+		}
+	}
```
Comment (aanm) [inline pkg/k8s/watchers/pod.go]: Shouldn't we make the assignment of the `*PodReady` variables conditional? The way the code is written right now it will make the pod to be considered "Ready" even if the first condition was not "Ready" and the last "Condition" was "Ready".

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/slim/k8s/apis/core/v1/types.go (hunk 1 above)
Comment (Weil0ng) [reply pkg/k8s/slim/k8s/apis/core/v1/types.go]: will do, but curious why?

File: pkg/k8s/watchers/pod.go (hunk 2 above)
Comment (Weil0ng) [reply pkg/k8s/watchers/pod.go]: I'm a bit confused, the statues variables (`oldPodReady` and `newPodReady`) are already conditioned on the *type* of the condition we are checking, not sure why other conditions would set them?

File: pkg/k8s/watchers/pod.go [hunk 3]
```diff
@@ -276,7 +277,24 @@ func (k *K8sWatcher) updateK8sPodV1(oldK8sPod, newK8sPod *slim_corev1.Pod) error
...
+	if lrpNeedsReassign || (oldPodReady != newPodReady) {
+		k.redirectPolicyManager.OnUpdatePod(newK8sPod, lrpNeedsReassign, newPodReady == slim_corev1.ConditionTrue)
```
Comment (aditighag) [inline pkg/k8s/watchers/pod.go:287]: Why do we need reassign? ~~Any reason why we can't simply call `OnDeletePod` here? There is already logic in place to remove the service entry if there are no backends.~~ 
Checked the code again, and looks like the pod itself doesn't get deleted, just the `PodCondition` changes between Ready/NotReady/etc?

**Review by aditighag — CHANGES_REQUESTED** [review]

File: pkg/k8s/watchers/pod.go (hunk 3 above)
Comment (aditighag) [reply pkg/k8s/watchers/pod.go:287]: I'm not able to post this comment at the exact code location since it's not part of your change, but we need to check the `podCondition` in `getLocalPodsForPolicy` in manager.go.

Comment (Weil0ng) [reply pkg/k8s/watchers/pod.go:287]: > looks like the pod itself doesn't get deleted, just the PodCondition changes between Ready/NotReady/etc?

yup, that's what this tries to capture here, on pod delete there should be another event that goes to `OnDeletePod`

> we need to check the podCondition in getLocalPodsForPolicy in manager.go

Makes sense, this is for 'Add' operation correct?

Comment (Weil0ng) [reply pkg/k8s/watchers/pod.go:287]: Done.

File: pkg/k8s/slim/k8s/apis/core/v1/types.go (hunk 1 above)
Comment (aanm) [reply pkg/k8s/slim/k8s/apis/core/v1/types.go]: @Weil0ng just to make it in the same order as the upstream version. It is easier to do a diff between both files if things are kept in the same order.

File: pkg/k8s/watchers/pod.go (hunk 2 above)
Comment (aanm) [reply pkg/k8s/watchers/pod.go]: @Weil0ng oh, I think I missed the `cond.Type == slim_corev1.PodReady`, so we are expecting that there's a single condition of type PodReady? If yes, couldn't we have a "break" on lines 288 and 294?

File: pkg/redirectpolicy/manager.go [hunk 4]
```diff
@@ -499,6 +509,16 @@ func (rpm *Manager) upsertService(config *LRPConfig, frontendMapping *feMapping)
 	}
 }
 
+func (rpm *Manager) checkPodReady(pod *slimcorev1.Pod) bool {
+	ready := slimcorev1.ConditionUnknown
+	for _, cond := pod.Status.Conditions {
```
Comment (pchaigno) [inline pkg/redirectpolicy/manager.go]: Compilation fails on this line with
```
../pkg/redirectpolicy/manager.go:514:39: syntax error: _, cond := pod.Status.Conditions used as value
```
I think you need:
```suggestion
	for _, cond := range pod.Status.Conditions {
```

**Review by pchaigno — CHANGES_REQUESTED** [review]

File: pkg/k8s/slim/k8s/apis/core/v1/types.go (hunk 1 above)
Comment (Weil0ng) [reply pkg/k8s/slim/k8s/apis/core/v1/types.go]: I see, thanks!

Comment (Weil0ng) [reply pkg/k8s/slim/k8s/apis/core/v1/types.go]: and done :)

File: pkg/k8s/watchers/pod.go (hunk 2 above)
Comment (Weil0ng) [reply pkg/k8s/watchers/pod.go]: yes, done.

File: pkg/redirectpolicy/manager.go (hunk 4 above)
Comment (Weil0ng) [reply pkg/redirectpolicy/manager.go]: oops

**Comment by Weil0ng** [comment]
test-me-please

**Comment by Weil0ng** [comment]
All tests are passing, PTAL :)

File: pkg/redirectpolicy/manager.go [hunk 5]
```diff
@@ -512,6 +532,9 @@ func (rpm *Manager) getLocalPodsForPolicy(config *LRPConfig, podStore cache.Stor
 		if err != nil {
 			continue
 		}
+		if !rpm.checkPodReady(pod) {
```
Comment (aditighag) [inline pkg/redirectpolicy/manager.go]: I don't think this will work as is. If you return here without storing the pods in the `rpm.policyPods` map, you'll never process future update events for the pod? I'll need to think about this a bit more, and I can push a stacked commit to your PR or if you want to give it a shot, feel free to do it.

Comment (Weil0ng) [reply pkg/redirectpolicy/manager.go]: why wouldn't we process update events for the pod? we should be able to detect a ready status change and call `OnUpdatePod` and insert it there, no?

Comment (aditighag) [reply pkg/redirectpolicy/manager.go]: I checked the code again, we iterate over all the current policies to see if a pod is selected. I think this should be fine. But please test the case I mentioned offline, where a pod is deployed before an LRP.

**Review by aditighag — APPROVED** [review]
LGTM. This is probably not easy to manually test, right? We should add some unit tests in follow-up PRs.

File: pkg/redirectpolicy/manager.go [hunk 6]
```diff
@@ -499,6 +509,16 @@ func (rpm *Manager) upsertService(config *LRPConfig, frontendMapping *feMapping)
 	}
 }
 
+func (rpm *Manager) checkPodReady(pod *slimcorev1.Pod) bool {
+	ready := slimcorev1.ConditionUnknown
+	for _, cond := range pod.Status.Conditions {
+		if cond.Type == slimcorev1.PodReady {
+			ready = cond.Status
```
Comment (christarazi) [inline pkg/redirectpolicy/manager.go]: Is there supposed to be a `break` after hitting this?

Comment (christarazi) [reply pkg/redirectpolicy/manager.go]: If I'm reading this right, it will set `ready` when the pod's condition is `PodReady`, but it continues looping.

File: pkg/k8s/watchers/pod.go [hunk 7]
```diff
@@ -276,7 +277,26 @@ func (k *K8sWatcher) updateK8sPodV1(oldK8sPod, newK8sPod *slim_corev1.Pod) error
...
+			oldPodReady = cond.Status
+			break
+		}
+	}
+	for _, cond := range newK8sPod.Status.Conditions {
+		if cond.Type == slim_corev1.PodReady {
+			newPodReady = cond.Status
+			break
+		}
+	}
```
Comment (christarazi) [inline pkg/k8s/watchers/pod.go]: Can define a helper function to do this? `getLatestPodCondition` or something like that. And then you could reuse it in `(rpm *Manager) checkPodReady`

**Review by christarazi — CHANGES_REQUESTED** [review]
Some items to address below. I also would like to see some unit tests. If we had tests, I think we would have been able to catch some of the bugs like missing a `break` and so on.

**Review by aanm — APPROVED** [review]
Nothing else besides what Chris point it out.

File: pkg/redirectpolicy/manager.go (hunk 6 above)
Comment (Weil0ng) [reply pkg/redirectpolicy/manager.go]: you are absolutely right, fixed.

File: pkg/k8s/watchers/pod.go (hunk 7 above)
Comment (Weil0ng) [reply pkg/k8s/watchers/pod.go]: Done.

**Comment by Weil0ng** [comment]
> LGTM. This is probably not easy to manually test, right? We should add some unit tests in follow-up PRs.

I've manually verified that this works after rebasing :) will add unit/e2e tests as a follow up PR if that's okay.

**Comment by Weil0ng** [comment]
test-me-please

**Comment by Weil0ng** [comment]
test-me-please

**Comment by Weil0ng** [comment]
PTAL :)

**Review by christarazi — APPROVED** [review]

File: pkg/redirectpolicy/manager.go [hunk 8]
```diff
@@ -492,6 +502,10 @@ func (rpm *Manager) upsertService(config *LRPConfig, frontendMapping *feMapping)
 	}
 }
 
+func (rpm *Manager) checkPodReady(pod *slimcorev1.Pod) bool {
```
Comment (aditighag) [inline pkg/redirectpolicy/manager.go]: nit: This method isn't required? Also, why is this added as a method receiver?

File: pkg/k8s/utils/utils.go [hunk 9]
```diff
@@ -109,6 +109,16 @@ func GetServiceListOptionsModifier() (func(options *v1meta.ListOptions), error)
 	}, nil
 }
 
+// GetLatestPodReadiness returns the lastest podReady condition on a given pod.
+func GetLatestPodReadiness(podStatus slim_corev1.PodStatus) slim_corev1.ConditionStatus {
+	for _, cond := range podStatus.Conditions {
+		if cond.Type == slim_corev1.PodReady {
```
Comment (aditighag) [inline pkg/k8s/utils/utils.go:115]: If I understand this correctly, we'll return `PodReady` as soon as we find one of the containers in the pod as ready? Is that the desired behavior? Why don't we check for `ContainersReady` type instead?

```
// ContainersReady indicates whether all containers in the pod are ready.
ContainersReady PodConditionType = "ContainersReady"
```

**Review by aditighag — CHANGES_REQUESTED** [review]

File: pkg/redirectpolicy/manager.go (hunk 8 above)
Comment (Weil0ng) [reply pkg/redirectpolicy/manager.go]: removed.

File: pkg/k8s/utils/utils.go (hunk 9 above)
Comment (Weil0ng) [reply pkg/k8s/utils/utils.go:115]: `PodReady` and `ContainerReady` are two different conditions (https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions), the former is more definitive in our context (that the pod is ready for traffic), so we only return the status of `PodReady` condition here.

Comment (aditighag) [reply pkg/k8s/utils/utils.go:115]: Also, the description for `PodReady` is as follows - 
```
	// PodReady means the pod is able to service requests and should be added to the
	// load balancing pools of all matching services.
	PodReady PodConditionType = "Ready"
```

Based on this, I wonder if the `PodReady` type is meant to be used in the endpoints/autoscaling context. I'm not able to find any clear documentation around which is to be used when.

Comment (aditighag) [reply pkg/k8s/utils/utils.go:115]: > `PodReady` and `ContainerReady` are two different conditions (https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions), the former is more definitive in our context (that the pod is ready for traffic), so we only return the status of `PodReady` condition here.

Yes, I understand that they are two different conditions, but this explanation is not too convincing for me. If `ContainerReady` is subsumed under `PodReady`, then it makes sense to use the latter. 

>that the pod is ready for traffic

What does this mean? Does this refer to autoscaling like requirements?

Comment (Weil0ng) [reply pkg/k8s/utils/utils.go:115]: >  If ContainerReady is subsumed under PodReady, then it makes sense to use the latter.

yup, my understanding is that `PodReady` = `ContainerReady` + readiness probe reporting success.

**Review by aditighag — APPROVED** [review]

**Comment by aditighag** [comment]
test-me-please

**Comment by Weil0ng** [comment]
All tests were passing on ** 31ba5a6**

Here's the diff with lasted commit:
```
cuiwl@cuiwl:~/go/src/github.com/cilium/cilium/pkg$ git diff 31ba5a6 9532d82
diff --git a/pkg/redirectpolicy/manager.go b/pkg/redirectpolicy/manager.go
index 0d87a8d9d..5a24fd87a 100644
--- a/pkg/redirectpolicy/manager.go
+++ b/pkg/redirectpolicy/manager.go
@@ -502,10 +502,6 @@ func (rpm *Manager) upsertService(config *LRPConfig, frontendMapping *feMapping)
        }
 }
 
-func (rpm *Manager) checkPodReady(pod *slimcorev1.Pod) bool {
-       return k8sUtils.GetLatestPodReadiness(pod.Status) == slimcorev1.ConditionTrue
-}
-
 // Returns a slice of endpoint pods metadata that are selected by the given policy config.
 func (rpm *Manager) getLocalPodsForPolicy(config *LRPConfig, podStore cache.Store) []*podMetadata {
        var (
@@ -529,7 +525,7 @@ func (rpm *Manager) getLocalPodsForPolicy(config *LRPConfig, podStore cache.Stor
                        }).Error("failed to get valid pod metadata")
                        continue
                }
-               if !rpm.checkPodReady(pod) {
+               if k8sUtils.GetLatestPodReadiness(pod.Status) != slimcorev1.ConditionTrue {
                        continue
                }
                if !config.policyConfigSelectsPod(podData) {
```
